### PR TITLE
Add API to upload translated files as pretranslation

### LINKED ISSUE #4478: Create new API to submit pretranslations
This should be implemented after #4447.

Using the same logic as upload translations, we should have an API to upload a file and store translations as pretranslations.

This API requires a PAT. Only users part of the manually curated `pretranslators` group can use it.

Behavior: 
- If a translation in the file matches a translated string, skip the entity.
- If an entity already has a pretranslation and the new translation is different, override it (reject previous pretranslation, store new translation as pretranslation).
- If an entity already has unreviewed suggestions, it should be pretranslated — unlike it happens with the standard pretranslation. If an existing suggestion matches the translation in the uploaded file, that suggestion should be approved (treating the match as a signal).

### PR DESCRIPTION
**Pull request by flodolo**
Fixes #4478 

Adds POST `/api/v2/upload/pretranslations/`, storing translations from an uploaded file as pretranslations. Requires membership of the `pretranslators` group on top of translator rights.

Approved translations are never touched. A matching suggestion or fuzzy translation is made the active pretranslation, keeping its original author; a different pretranslation or fuzzy translation is rejected and replaced.

### CODE REVIEW

**Comment by flodolo** [comment]
This turned out way more complicated than I expected, mostly because I kept finding edge cases (first with fuzzy, then with race conditions, finally with DB vs non-DB checks). Hence the amount of tests 😵 

- `pontoon/sync/core/translations_from_repo.py` is refactor only. Extracted `build_translation()`, which both the existing sync path and the new import now use.
- Most of `pontoon/api/views.py` is the UploadView base class, factored out of the existing translations endpoint: shared request validation, transaction/error handling, and the truncated key lists. The pretranslations view itself is small.
- UNDEFINED_KEYS_LIMIT → UPLOAD_KEYS_ERROR_LIMIT in `serializers.py`, since it now truncates two lists.
- Uploads marked fuzzy are skipped, and anything failing a quality check is left out rather than replacing a good translation. The check result comes from `run_checks()` instead of the stored Error/Warning rows, so it also catches checks Pontoon doesn't persist — Android placeholder mismatches, for example, until #4499 if fixed. I think it's OK for this endpoint to be strict.

File: pontoon/sync/utils.py [hunk 1]
```diff
@@ -150,3 +180,257 @@ def import_uploaded_file(
...
+    )
+
+    current: dict[int, list[Translation]] = defaultdict(list)
+    for tx in Translation.objects.filter(
+        entity__resource=db_res,
+        entity__obsolete=False,
+        locale=locale,
+        rejected=False,
+    ).iterator():
+        current[tx.entity_id].append(tx)
```
Comment (eemeli) [inline pontoon/sync/utils.py]: This looks like it should be using `itertools.groupby`.

File: pontoon/sync/utils.py [hunk 2]
```diff
@@ -150,3 +180,257 @@ def import_uploaded_file(
...
+        return self.new or self.match
+
+
+def import_uploaded_pretranslations(
+    project: Project,
+    locale: Locale,
+    db_res: DbResource,
+    upload: File,
+    user: User,
+) -> PretranslationUploadResult:
```
Comment (eemeli) [inline pontoon/sync/utils.py:244]: This function seems to implement much of the same work as is done in `pontoon.sync.core.translations_from_repo.update_db_translations`, but does so differently enough that it's hard to say _how_ they differ.

Is the work done for pretranslations different enough that it really needs its own separate function? And is this function extensible to also support uploads as suggestions, or will that need a third function?

**Review by eemeli — COMMENTED** [review]
Not going to block as I don't have the bandwidth to review thoroughly and iterate before I go on PTO, but I'm concerned about the number of new classes and functions that are added here. At the very least we should establish `pontoon.sync.upload` as a place for them, but also make sure that they're all actually useful and needed.

File: pontoon/sync/utils.py (hunk 1 above)
Comment (flodolo) [reply pontoon/sync/utils.py]: I'll make the change, but it's not a straightforward swap (need to order the rows here, and make sure there is a default later).

File: pontoon/sync/utils.py (hunk 2 above)
Comment (flodolo) [reply pontoon/sync/utils.py:244]: Fair question. The two functions disagree on almost every decision:

 - Approved translations: sync overrides them (repo wins); here the entity is skipped, a review is never undone.
 - Match criterion: sync matches on string inside the DB query; here it's `translations_equal()` on value + properties.
 - Non-matching rows: sync rejects everything else; here only a previous pretranslation or fuzzy translation is rejected, while plain suggestions are just deactivated and kept.
 - Fuzzy in the upload: stored as fuzzy by sync, skipped here.
 - Checks: sync runs them after writing (bulk_run_checks); here they run on the staged translations before anything is written, so a failing upload can't discard the translation it would have replaced.
 - TM entries: created by sync, not here (they will be created when pretranslations are approved).
 - Output: log lines vs. per-key counters and failed-check details for the API response.
 
Folding this logic into `update_db_translations` means a mode flag at 7–8 branch points. IMO that's much harder to follow, so I kept them separate and reused what's shared (`parse_upload_for_entities`, `translations_equal`, `build_translation`).

 "Upload as suggestions" should be closer in shape to this function: don't touch approved translations, match by value, keep the match's author. The difference is: not setting pretranslated and not rejecting or deactivating anything. A mode parameter in this function should be enough, but I don't have an implementation for that yet.

Should I spell the sync-vs-upload differences out in the docstring? Maybe a table with the choices (sync, pretranslation, suggestion) would be easier to parse that text.

Comment (eemeli) [reply pontoon/sync/utils.py:244]: > Match criterion: sync matches on string inside the DB query; here it's `translations_equal()` on value + properties.

This should be fixed in `update_db_translations`, as I mentioned in https://github.com/mozilla/pontoon/pull/4491#discussion_r3948628309.

> Fuzzy in the upload: stored as fuzzy by sync, skipped here.

Why? If a pretranslation upload includes something newly marked as fuzzy, why skip it? Conceptually, "pretranslated" and "fuzzy" are _really_ similar statuses for a translation, so I'd rather expect for one of the following:
- The flag is retained on the stored pretranslation.
- The flag is ignored, and the pretranslation stored without it.
- This is considered an error.

Silently skipping the pretranslation seems like a worse option than any of the above, as it does a poorer job of communicating expectations to the user or accommodating their communicated intent.

> Should I spell the sync-vs-upload differences out in the docstring? Maybe a table with the choices (sync, pretranslation, suggestion) would be easier to parse that text.

My main issue here is that the code is architected completely differently between sync and upload, but they do very similar things. This makes it significantly more difficult to get an understanding of the two implementations, and figure out their differences. So if/when we need to make any changes to sync code, which is already plenty complex, we need to minimize the mental load we impose and not implement the same thing in more than one way.

If the code style used in this PR's `import_uploaded_translations` seems better and more readable than what we're now doing to map the differences between what we've in the database and what's in the upload/sync, then that style should be used for the sync as well -- or vice versa.

Comment (flodolo) [reply pontoon/sync/utils.py:244]: > Why? If a pretranslation upload includes something newly marked as fuzzy, why skip it? Conceptually, "pretranslated" and "fuzzy" are _really_ similar statuses for a translation, so I'd rather expect for one of the following:
> 
>     * The flag is retained on the stored pretranslation.
>     * The flag is ignored, and the pretranslation stored without it.
>     * This is considered an error.

A fuzzy string is generated by gettext automation. Depending on how the developer set it, the fuzzy string could be completely unrelated and wrong (that's why we asked devs at Mozilla to disable it). 

If a string is fuzzy, it shouldn't end up in the repo and being used. The person uploading can remove the fuzzy flag from their file if they think that's ready to be used.

> If the code style used in this PR's `import_uploaded_translations` seems better and more readable than what we're now doing to map the differences between what we've in the database and what's in the upload/sync, then that style should be used for the sync as well -- or vice versa.

I'm happy to file this as a follow up once this API work is complete. I personally don't want to touch the main sync code.

File: pontoon/api/README.md [hunk 3]
```diff
@@ -156,5 +156,77 @@ Status codes:
...
+  ],
+  "failed_checks_count": 1,
+  "undefined_keys": [["obsolete_key"]],
+  "undefined_keys_count": 1
+}
+```
+
+- `created`: pretranslations added for strings with no pretranslation or fuzzy
+  translation.
+- `updated`: pretranslations replacing a previous, different pretranslation or fuzzy
```
Comment (mathjazz) [inline pontoon/api/README.md]: _Converted_ are also _updated_. Maybe _replaced_?

File: pontoon/api/README.md [hunk 4]
```diff
@@ -156,5 +156,77 @@ Status codes:
...
+  first 100 keys are listed.
+- `failed_checks_count`: total number of strings left untouched because of failing
+  checks, before truncation.
+- `undefined_keys`: keys of translations with no matching string in Pontoon, ignored.
+  Each key is a list of strings, in the same format as the `key` field of entities.
+  Only the first 100 keys are listed.
+- `undefined_keys_count`: total number of keys with no matching string in Pontoon,
+  before truncation.
+
+Unlike the built-in pretranslation, strings with unreviewed suggestions are
```
Comment (mathjazz) [inline pontoon/api/README.md:221]: Shall we file an issue to change the behaviour for built-in pretranslations (or even fix that as part of this PR)?

File: pontoon/api/README.md [hunk 5]
```diff
@@ -156,5 +156,77 @@ Status codes:
...
+pretranslated. Suggestions that don't match the uploaded translation are kept as
+unreviewed suggestions. Fuzzy translations already in Pontoon are rejected and replaced
+by the pretranslation, while approved translations are never replaced. An uploaded
+translation that fails any quality check is neither stored nor converted, and the
+previous translation is left in place, so that a broken translation never replaces a
+good one. The checks are the ones the editor runs, including those Pontoon does not
+store in the database.
```
Comment (mathjazz) [inline pontoon/api/README.md]: The editor also runs Translate Toolkit checks by default.

Same question as above, shall we file an issue to unify the checks behaviour between the API and the built-in pretranslations (or even fix that as part of this PR)?

File: pontoon/api/views.py [hunk 6]
```diff
@@ -629,51 +631,22 @@ def post(self, request):
 class UploadConflict(APIException):
     status_code = status.HTTP_409_CONFLICT
     default_detail = (
-        "A concurrent upload changed the same translations. Retry the request."
+        "A concurrent upload or review changed the same translations. "
```
Comment (mathjazz) [inline pontoon/api/views.py:634]: Why _or review_?

File: pontoon/api/README.md [hunk 7]
```diff
@@ -156,5 +156,77 @@ Status codes:
...
+    { "key": ["entity_key"], "errors": ["Double space"], "warnings": [] }
+  ],
+  "failed_checks_count": 1,
+  "undefined_keys": [["obsolete_key"]],
+  "undefined_keys_count": 1
+}
+```
+
+- `created`: pretranslations added for strings with no pretranslation or fuzzy
+  translation.
```
Comment (mathjazz) [inline pontoon/api/README.md]:
```suggestion
- `created`: number of pretranslations added for strings with no pretranslation or fuzzy
  translation.
```

Also applies to keys below.

File: pontoon/sync/utils.py
Comment (mathjazz) [inline pontoon/sync/utils.py:1]: This module has outgrown its name - most of the top-level definitions are upload-specific. I'd move them to upload.py.

File: pontoon/sync/utils.py [hunk 8]
```diff
@@ -150,3 +180,274 @@ def import_uploaded_file(
...
+    entities_by_id = {
+        entity.pk: entity
+        for entity in Entity.objects.filter(
+            pk__in={p.translation.entity_id for p in pending}
+        )
+    }
+    applied: list[_PendingPretranslation] = []
+    for p in pending:
+        entity = entities_by_id[p.translation.entity_id]
+        entity.resource = db_res
```
Comment (mathjazz) [inline pontoon/sync/utils.py]: For DTD resources `compare_locales` evaluates `entity.resource.entities.all()`, reloading every entity of the resource on each call. We should prefetch.

File: pontoon/sync/utils.py [hunk 9]
```diff
@@ -150,3 +180,274 @@ def import_uploaded_file(
...
+        ActionLog.objects.bulk_create(actions)
+
+    if new_translations:
+        # bulk_create() skips Translation.save(), which would do this
+        new_translations[0].update_latest_translation()
+
+    changed_pks = [tx.pk for tx in new_translations + converted_translations]
+    if changed_pks or reject_ids:
+        update_stats(project)
+        Translation.objects.filter(pk__in=reject_ids + changed_pks).bulk_mark_changed()
```
Comment (mathjazz) [inline pontoon/sync/utils.py]: `bulk_mark_changed` calls `bulk_creates` without `ignore_conflicts` against a `unique_together`. If another submission happens in the editor at the same time, it will raise `IntegrityError` and discards the entire upload. 

Above we deliberately use `bulk_create(ignore_conflicts=True)` instead and I suggest we do so here as well or add it to `bulk_mark_changed`.

File: pontoon/api/views.py (hunk 6 above)
Comment (flodolo) [reply pontoon/api/views.py:634]: A string with a pending suggestion would get replaced by pretranslation. If a person approved it in the meantime (as part of review), we would end up with an integrity error.

File: pontoon/api/README.md (hunk 4 above)
Comment (flodolo) [reply pontoon/api/README.md:221]: I would do both this and the TT checks in a follow-up. I think it makes sense to pretranslate strings with suggestions in both, we need to make a decision on TT checks (I think we should ignore them, since they're unreliable), but both should be documented in the issue.

**Comment by flodolo** [comment]
Kept the code changes and code reorganization in 2 separate commits.

File: pontoon/api/README.md (hunk 4 above)
Comment (mathjazz) [reply pontoon/api/README.md:221]: Sounds good. 

TT checks should remain disabled in both, so I'd drop the "checks are the ones the editor runs" reference below. I believe the only difference are the `XCODE`, `GETTEXT` and `WEBEXT` checks in `custom.py`.

File: pontoon/api/views.py (hunk 6 above)
Comment (mathjazz) [reply pontoon/api/views.py:634]: I assume if upload and translation creation happen at the same time we hit the same error?

File: pontoon/sync/upload.py [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+    if deactivate_ids:
+        Translation.objects.filter(pk__in=deactivate_ids).update(active=False)
+    if converted_translations:
+        Translation.objects.bulk_update(
+            converted_translations, ["active", "fuzzy", "pretranslated"]
+        )
+    if new_translations:
+        Translation.objects.bulk_create(new_translations)
+
+    # A review may approve one of these translations after `current` was loaded.
```
Comment (mathjazz) [inline pontoon/sync/upload.py]: Sorry, missed this one earlier: What if the review is a rejection?

The `bulk_update above doesn't write `rejected`, so a rejection landing during the checks loop leaves a `rejected=True`, `pretranslated=True`, `active=True` row. The likelihood of that is low, though.

**Review by mathjazz — COMMENTED** [review]
Thanks for the update!

File: pontoon/api/views.py (hunk 6 above)
Comment (flodolo) [reply pontoon/api/views.py:634]: Yes. Creating a translation in the editor for the same entity and locale makes it active, which collides with the `entity_locale_active` constraint when the upload activates its own translation. That raises an `IntegrityError`, which is caught and returned as a 409 as well.

File: pontoon/sync/upload.py (hunk 10 above)
Comment (flodolo) [reply pontoon/sync/upload.py]: Good point, it should be possible to catch that.

**Review by mathjazz — APPROVED** [review]
Looks good, great work!

Please file the followup issues.
